presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetCompressUtils.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.parquet;
+
+import io.airlift.slice.Slice;
+import org.iq80.snappy.Snappy;
+import parquet.hadoop.metadata.CompressionCodecName;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.zip.GZIPInputStream;
+
+import static io.airlift.slice.Slices.EMPTY_SLICE;
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static parquet.hadoop.metadata.CompressionCodecName.GZIP;
+import static parquet.hadoop.metadata.CompressionCodecName.SNAPPY;
+import static parquet.hadoop.metadata.CompressionCodecName.UNCOMPRESSED;
+import static sun.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
+
+public class ParquetCompressUtils
+{
+    private ParquetCompressUtils() {}
+
+    public static Slice decompress(CompressionCodecName codec, Slice input, int uncompressedSize)
+            throws IOException
+    {
+        if (input.length() == 0) {
+            return EMPTY_SLICE;
+        }
+        switch (codec) {
+            case GZIP:
+                return decompressGzip(input, uncompressedSize);
+            case SNAPPY:
+                return decompressSnappy(input, uncompressedSize);
+            case UNCOMPRESSED:
+                return wrappedBuffer(input.getBytes());
+            default:
+                // TODO add LZO support using pure java aircompressor
+                throw new ParquetCorruptionException("Parquet not support codec: " + codec);
+        }
+    }
+
+    private static Slice decompressSnappy(Slice input, int uncompressedSize)
+    {
+        byte[] inArray = (byte[]) input.getBase();
+        int inOffset = (int) (input.getAddress() - ARRAY_BYTE_BASE_OFFSET);
+        int inLength = input.length();
+
+        byte[] buffer = new byte[uncompressedSize];
+        Snappy.uncompress(inArray, inOffset, inLength, buffer, 0);
+        return wrappedBuffer(buffer);
+    }
+
+    private static Slice decompressGzip(Slice input, int uncompressedSize)
+            throws IOException
+    {
+        // TODO use java zip inflater
+        ByteArrayOutputStream uncompressedStream = new ByteArrayOutputStream();
+        InputStream gzipInputStream = new GZIPInputStream(new ByteArrayInputStream(input.getBytes()));
```

Comment:
You may want to close this stream as it may leak native memory (see [this](http://www.evanjones.ca/java-native-leak-bug.html)).